replicate-to-sql-with-generic-events/lib/index.js

```diff
@@ -0,0 +1,60 @@
+const { PostgresWrapper, pgConfigDocker } = require('./postgres');
+
+class CorePlugin {
+  constructor() {
+    this.context = null;
+    this.config = {};
+    this.pipes = {
+      'generic:document:afterWrite': 'afterWrite',
```

Comment:
There is fundamentally no solution to this problem as long as you consider that these 2 systems can have different validation conditions, because in that case, you'll have to rollback manually. This is desastrous from a performances point of view, and this can lead to undesirable side-effects because unwanted data will be present in one of the 2 systems, even for a short while. And let's not mention the case when the rollback fails.

For this to work properly, you have to chose one of the following strategies:

# Strategies 

## the 2 storages have the exact same set of restrictions

In that case, I would keep the code in the "after" event, as it will only be invoked if the insertion in ES is successful. Job's done with 0 lines of code written

## storage A is more restrictive than storage B, and storage B's restrictions are a subset of storage A's ones

In that case, data must be inserted first in storage A. If this succeeds, then you're guaranteed to be able to insert them in storage B as well.

This is usually the case when you duplicate the storage between a NoSQL and a relational database. In that case, the relational db is usually the most restrictive storage of the two : in that case, I would insert first in postgres in a "before" action : in case of a failure, postgres will handle the rollback itself, all we have to do is to throw an error in the pipe function to instruct Kuzzle to not pursue with storing the data in ES.

# Conclusions

1. there is no "best" solution, this all depends on what you need to do and what your datamodel is.

2. I still don't think that an "around" pipe is a valid solution in that case, for me it leads to even more problems

Reply:
@Aschen @rolljee 

There is fundamentally no solution to this problem as long as you consider that these 2 systems can have different validation conditions, because in that case, you'll have to rollback manually. This is desastrous from a performances point of view, and this can lead to undesirable side-effects because unwanted data will be present in one of the 2 systems, even for a short while. And let's not mention the case when the rollback fails.

For this to work properly, you have to chose one of the following strategies:

# Strategies 

## the 2 storages have the exact same set of restrictions

In that case, I would keep the code in the "after" event, as it will only be invoked if the insertion in ES is successful. Job's done with 0 lines of code written

## storage A is more restrictive than storage B, and storage B's restrictions are a subset of storage A's ones

In that case, data must be inserted first in storage A. If this succeeds, then you're guaranteed to be able to insert them in storage B as well.

This is usually the case when you duplicate the storage between a NoSQL and a relational database. In that case, the relational db is usually the most restrictive storage of the two : in that case, I would insert first in postgres in a "before" action : in case of a failure, postgres will handle the rollback itself, all we have to do is to throw an error in the pipe function to instruct Kuzzle to not pursue with storing the data in ES.

# Conclusions

1. there is no "best" solution, this all depends on what you need to do and what your datamodel is.

2. I still don't think that an "around" pipe is a valid solution in that case, for me it leads to even more problems